ts/private/ts_project.bzl

```diff
@@ -182,83 +182,111 @@ See https://github.com/aspect-build/rules_ts/issues/361 for more details.
         fail(transpiler_selection_required)
 
     output_types = typings_outs + typing_maps_outs + typings_srcs
-    use_tsc_noemit = ctx.attr.no_emit or (ctx.attr.transpile == 0 and not ctx.attr.declaration)
+
+    # What tsc will be emitting
+    use_tsc_for_js = len(js_outs) > 0
+    use_tsc_for_dts = len(typings_outs) > 0
+
+    # Use a separate non-emitting action for type-checking when:
+    #  - a separate action was explicitly requested
+    #  or
+    #  - not invoking tsc for output files at all
+    use_tsc_noemit = ctx.attr.isolated_typecheck or not (use_tsc_for_js or use_tsc_for_dts)
```

Comment:
nit: may as well call this `use_isolated_typecheck`. `--noEmit` is implied on that action.